Add generic device and initial support in portBLAS

s-Nick opened the pull request:
# Description

 - oneMKL Interfaces currently only supports known targets: Intel CPU/GPU, AMD GPU, Nvidia GPU
- This PR:
    - Enables a new generic target
    - Enables the generic target to use the portBLAS backend
    - Adds documentation
    - Relates to https://github.com/oneapi-src/oneMKL/issues/542

# Checklist

## All Submissions

- [ ] Do all unit tests pass locally? Attach a log.
  - I don't have a non Intel/NVIDIA/AMD device to test with. Tests are successful when forcing Intel devices to go through `generic_device` path.

Following logs show that all domains still work properly:
[arc_mkl_BLAS_log.txt](https://github.com/user-attachments/files/16886641/arc_mkl_BLAS_log.txt)
[arc_mkl_DFT_log.txt](https://github.com/user-attachments/files/16886644/arc_mkl_DFT_log.txt)
[arc_mkl_lapack_log.txt](https://github.com/user-attachments/files/16886646/arc_mkl_lapack_log.txt)
[arc_mkl_rng_log.txt](https://github.com/user-attachments/files/16886648/arc_mkl_rng_log.txt)


- [x] Have you formatted the code using clang-format?


## New features

- [x] Have you provided motivation for adding a new feature?
- [x] Have you added relevant tests?
The generic device doesn't requires new tests.

hjabird committed:
Add generic device; Initial support in portBLAS

* oneMKL Interfaces currently only supports known targets: Intel CPU/GPU, AMD GPU, Nvidia GPU
* This PR:
  * Enables a new generic target
  * Enables the generic target to use the portBLAS backend
  * Adds documentation

hjabird committed:
CMake typo; Add -fno-sycl-instrument-device-code

s-Nick committed:
remove ENABLE_GENERIC_DEVICE option and add always generic_device to
possible devices.

This commit remove the option ENABLE_GENERIC_DEVICE and instead add generic_device
to the backends_table. The check for unsupported_device exception is moved to
table_initializer and to keep it as informative as it is, it is required a
change to the function_tables operator[]. This change allows to use
portBLAS (and in a possible future) all "port" libraries with  any
device supported.

s-Nick committed:
Enable test for OpenCL GPUs except for Intel's one

This patch enables the possibility to run tests with generic_device for
devices that have an OpenCL backend.

s-Nick committed:
Add pragma to guard generic device usage and exception

s-Nick committed:
Fix typo

Rbiessy commented on src/include/function_table_initializer.hpp:
```diff
@@ -99,8 +99,17 @@ class table_initializer {
         if (!handle) {
-            std::cerr << ERROR_MSG << '\n';
-            throw mkl::backend_not_found();
+#ifndef ENABLE_PORTBLAS_BACKEND
```
I think it would look cleaner if this `ifndef` is separated to create a variable like
```cpp
#if defined(ENABLE_PORTBLAS_BACKEND)
static constexpr bool is_generic_device_supported = true;
#else
static constexpr bool is_generic_device_supported = false;
#endif
```
That will keep the runtime `if` cleaner.

s-Nick replied:
Thank you for the comment @Rbiessy I addressed it in [3466d6f](https://github.com/oneapi-src/oneMKL/pull/566/commits/3466d6f87bc5e0f720dd9a538cd491c5439fe893)

s-Nick committed:
Move generic_device support pragma

Moved pragma and simplified if-statement to increase code readability

s-Nick committed:
Fix broken if-statement

Rbiessy reviewed (approved):
That looks good to me, thanks!

s-Nick commented:
@oneapi-src/onemkl-maintain Hi all, in this PR I am adding a new generic device to allow portable libraries such as portBLAS and portFFT to be compiled and run on devices that are not known to us right now, but that have an OpenCL backend available. 
 
Changes are few and simple, but since I am making an important one to `table_initializer` and it touches all domains, I would like to have your approval on it. These changes don't cause any issue to current backends as you can check from logs attached.

andrewtbarker commented on tests/unit_tests/main_test.cpp:
```diff
@@ -101,15 +101,17 @@ int main(int argc, char** argv) {
-                                            "OpenCL") != std::string::npos)
+                    unsigned int vendor_id =
+                        static_cast<unsigned int>(dev.get_info<sycl::info::device::vendor_id>());
+                    /* Do not test for OpenCL backend on Intel GPU */
```
Does this logic continue to do what we want on AMD and nVidia GPUs?

s-Nick replied:
Thank you for your review @andrewtbarker 
As far as I know NVIDIA and AMD GPUs don't use OpenCL backends, so I expect they continue to work as intended. In any case, I am happy to add a condition to check if the vendor is one of the currently known one and skip the OpenCL backend.

Rbiessy replied:
> Does this logic continue to do what we want on AMD and nVidia GPUs?

What is the behavior that we want on AMD and Nvidia exactly? We've been wondering whether we should remove this check entirely since it's not clear why would we want to skip OpenCL backends. To me users should use `ONEAPI_DEVICE_SELECTOR` instead.

andrewtbarker commented on src/include/function_table_initializer.hpp:
```diff
@@ -59,14 +59,20 @@ class table_initializer {
 public:
-    function_table_t &operator[](oneapi::mkl::device key) {
-        auto lib = tables.find(key);
+    function_table_t &operator[](std::pair<oneapi::mkl::device, sycl::queue> device_queue_pair) {
```
It's not clear to me why this change is necessary. Where do we need the `sycl::queue` in the lookup tables?

s-Nick replied:
Since I moved the `unsupported_device` exception in the table initializer, I need the queue to query the device name. This is the only reason why I need the `sycl::queue` here.

andrewtbarker replied:
That makes sense. It seems like a large change for just the exception - I would support reverting the function table and throwing a more generic exception (or maybe implement `mkl::unsupported_device` with no device argument?). But it is also okay as is.

andrewtbarker reviewed (commented):
This is a useful change, I have just a couple questions.

s-Nick committed:
Update vector of device for testing

Previously if some specific backend where enabled the test suite always added a
cpu to the device to run test on, even if another if condition already
added them. This behaviour cause linking time issue if a cpu device is
not available.
This commit removes it and it adds missing pragma to
the other device selection, fixing the linking issue.

al3x-jp commented:
Hi, I got the following error when trying to build this checkout: 

[ 51%] Linking CXX executable ../../bin/test_main_blas_ct
terminate called after throwing an instance of 'sycl::_V1::runtime_error'
  what():  No device of requested type 'info::device_type::cpu' available. -1 (PI_ERROR_DEVICE_NOT_FOUND)
CMake Error at /usr/share/cmake-3.22/Modules/GoogleTestAddTests.cmake:83 (message):
  Error running test executable.
 
    Path: '/<path>/oneMKL/build/bin/test_main_blas_ct'
    Result: Subprocess aborted
    Output:
 
Call Stack (most recent call first):
  /usr/share/cmake-3.22/Modules/GoogleTestAddTests.cmake:179 (gtest_discover_tests_impl)
 
gmake[2]: *** [tests/unit_tests/CMakeFiles/test_main_blas_ct.dir/build.make:388: bin/test_main_blas_ct] Error 1

Does the build require a SYCL CPU target being available?
Thanks

s-Nick commented:
Hi @al3x-jp, thank you for highlighting this issue here. 
We were able to reproduce the same linking error using the `develop` branch and an open source DPC++ build.
In any case, commit [29e94fc](https://github.com/oneapi-src/oneMKL/pull/566/commits/29e94fcb56ce84e7f732bd0f8a63a3762a307611) addresses what we think caused it and on our machine and everything works fine. Could you please test again this PR including this last commit? Your feedback would be very helpful for us.

mkrainiuk commented on docs/building_the_project_with_dpcpp.rst:
```diff
@@ -408,6 +423,21 @@ set, the backend libraries to enable the use of BLAS, LAPACK and RNG with MKLGPU
+      -DENABLE_MKLGPU_BACKEND=False \
+      -DENABLE_PORTBLAS_BACKEND=True
+
+Note that this is not a supported configuration. This builds oneMKL Interfaces
```
Does it make sense to mention that we do not test it, but it's supported by portBLAS backend? To me "not supported" means we don't have backend for that, but we have, right?
```suggestion
Note that this is not a tested configuration. This builds oneMKL Interfaces
```

s-Nick replied:
Thank you for your review @mkrainiuk,
I agree with you, we have a backend for it (portBLAS) so tested it's better.

mkrainiuk commented on tests/unit_tests/main_test.cpp:
```diff
@@ -151,14 +153,6 @@ int main(int argc, char** argv) {
 #endif
     }
 
-#if defined(ENABLE_MKLCPU_BACKEND) || defined(ENABLE_NETLIB_BACKEND) || \
```
Could you please comment why we don't need this block anymore?

s-Nick replied:
This block always adds a CPU device for testing, but CPUs devices are added also in this [block of code](https://github.com/oneapi-src/oneMKL/blob/develop/tests/unit_tests/main_test.cpp#L113-L143). I was going to fix it only for portBLAS but it seems to me that it repeats the logic above so we don't need it.

mkrainiuk replied:
Got it, thank you.

s-Nick committed:
Update docs/building_the_project_with_dpcpp.rst

Co-authored-by: Maria Kraynyuk <maria.kraynyuk@intel.com>

lhuot reviewed (approved):
LGTM, thanks!
